src/legacy_handler.rs

```diff
@@ -93,73 +104,129 @@ impl LegacyHandler {
+
+        // Opt out of the global request/response handler as we implement an
+        // alternative flow here.
+        self.skip_request = true;
+
+        let mut file = OpenOptions::new().read(true).open(&args.image_path).await?;
+        let file_size = file.metadata().await?.len() as usize;
         let file_name = args
-            .file
+            .image_path
             .file_name()
-            .unwrap_or_default()
+            .ok_or(anyhow::anyhow!("file_name could not be extracted"))?
             .to_string_lossy()
             .to_string();
-        let bytes = tokio::fs::read(&args.file).await?;
-        let part = Part::stream(bytes)
-            .mime_str("application/octet-stream")?
-            .file_name(file_name);
-        Ok(Some(reqwest::multipart::Form::new().part("file", part)))
-    }
 
-    async fn handle_flash(&mut self, args: &FlashArgs) -> anyhow::Result<Option<Form>> {
-        let mut serializer = self.url.query_pairs_mut();
-        serializer
+        self.request
+            .url_mut()
+            .query_pairs_mut()
             .append_pair("opt", "set")
             .append_pair("type", "flash")
+            .append_pair("file", &file_name)
+            .append_pair("length", &file_size.to_string())
             .append_pair("node", &(args.node - 1).to_string());
 
-        if args.local {
-            serializer.append_pair("file", &args.image_path.to_string_lossy());
-            Ok(None)
-        } else {
-            println!("Warning: large files will very likely to fail to be uploaded in the current version");
-            let file_name = args
-                .image_path
-                .file_name()
-                .unwrap_or_default()
-                .to_string_lossy()
-                .to_string();
-            let bytes = tokio::fs::read(&args.image_path).await?;
-            let part = Part::stream(bytes)
-                .mime_str("application/octet-stream")?
-                .file_name(file_name);
-            Ok(Some(reqwest::multipart::Form::new().part("file", part)))
+        let response = Client::new()
+            .execute(self.request.try_clone().unwrap())
+            .await
+            .context("flash request")?;
+
+        if !response.status().is_success() {
+            bail!("could not execute flashing :{}", response.text().await?);
         }
+
+        let (sender, mut receiver) = channel::<bytes::Bytes>(256);
+
+        let read_task = async move {
+            let mut bytes_read = 0;
+            while bytes_read < file_size {
+                let read_len = BUFFER_SIZE.min(file_size - bytes_read);
+                let mut buffer = BytesMut::zeroed(read_len);
+                let read = file.read(&mut buffer).await?;
+                if 0 == read {
+                    // end_of_file
+                    break;
+                }
+                bytes_read += read;
+                buffer.truncate(read);
+                sender.send(buffer.into()).await?;
```

Comment:
Can't you just have 1 task, avoiding an MPSC queue?

```rust
while bytes_read < file_size {
    let buffer = file.read();
    request.send(buffer);
}
```

Reply:
I intend to implement an optimization. By decoupling and running both, reading the file, and sending the data in parallel, I aim to achieve the following:
* To spend as much time sending data over the TCP socket as possible.
* Smoothing out any hick-ups in reading or sending data by buffering the reads (in the channel). This comes with a minor memory penalty, which is reasonable given that our build targets are not memory-constrained as embedded micro's are.